### PR TITLE
[src] Add Fp16 grad sync for ddp

### PR DESCRIPTION
**Pull request by yuekaizhang**
Aishell test Accuracy with fp16_grad_sync enabled:
ctc greedy: 5.09
ctc prefix: 5.09
attention: 4.99
rescore: 4.69
It could improve training speed from 17 hours/100 epochs to 11 hours/100 epochs, especially for multi-nodes training.
My training environment:
two nodes with 6 gpus per node. eth0 Speed: 10000Mb/s

### CODE REVIEW

**Comment by yuekaizhang** [comment]
@whiteshirt0429 @robin1001

**Comment by robin1001** [comment]
Nice work. what about acc_grad, network card speed in our experiment?

**Comment by yuekaizhang** [comment]
> Nice work. what about acc_grad, network card speed in our experiment?

I used default train_conformer.yaml config and change the accu_grad to 1 (since I used 12 gpus total I believe the batch size is big enough) The network speed: sudo ethtool eth0 | grep Speed would give me 10000Mb/s

**Review by robin1001 — APPROVED** [review]

**Review by whiteshirt0429 — COMMENTED** [review]
+1

**Comment by xiehengye** [comment]
I use multi-nodes training.
My training  environment: two nodes with 4 gpus(p100) per node. eth0 Speed: 1000Mb/s.
I see no linear acceleration,Is it the network card? How to locate the problem

**Comment by xiehengye** [comment]
I use multi-nodes training.
My training environment: two nodes with 4 gpus(p100) per node. eth0 Speed: 1000Mb/s.
I see no linear acceleration,Is it the network card? How to locate the problem

**Comment by yuekaizhang** [comment]
> I use multi-nodes training. My training environment: two nodes with 4 gpus(p100) per node. eth0 Speed: 1000Mb/s. I see no linear acceleration,Is it the network card? How to locate the problem

It's hard to have linear acceleration. What's your current acceleration ratio? I think ~ 1.5 times with two nodes should be normal. Also, try to profile your training codes to check if the bottleneck is network card speed.

**Comment by xiehengye** [comment]
When I run training tasks, the maximum speed of the nic is 104 Mb/s,acceleration ratio is 1.18 times.I use a mechanical hard drive

**Comment by xiehengye** [comment]
Could you please tell me the machine configuration you used in the experiment? Thank you
